File: BukkitUtil-core/src/main/java/no/kh498/util/jackson/serializers/BukkitSerializers.java
package no.kh498.util.jackson.serializers;

import com.fasterxml.jackson.databind.BeanDescription;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializationConfig;
import com.fasterxml.jackson.databind.ser.Serializers;
import no.kh498.util.jackson.serializers.bean.ConfigurationSectionSerializer;
import no.kh498.util.jackson.serializers.bean.ConfigurationSerializableSerializer;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.serialization.ConfigurationSerializable;

/**
 * @author Elg
 */
public class BukkitSerializers extends Serializers.Base {

    @Override
    public JsonSerializer<?> findSerializer(SerializationConfig config, JavaType type, BeanDescription beanDesc) {
        if (ConfigurationSerializable.class.isAssignableFrom(type.getRawClass())) {
            return ConfigurationSerializableSerializer.inst;
        }
        if (ConfigurationSection.class.isAssignableFrom(type.getRawClass())) {
            return ConfigurationSectionSerializer.inst;
        }
        return null;
    }

}

File: BukkitUtil-core/src/main/kotlin/no/kh498/util/jackson/serializers/BukkitSerializers.kt
package no.kh498.util.jackson.serializers

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.*
import com.fasterxml.jackson.databind.ser.Serializers
import com.fasterxml.jackson.databind.ser.std.StdSerializer
import com.fasterxml.jackson.databind.type.MapType
import no.kh498.util.jackson.BukkitModule
import no.kh498.util.jackson.mixIn.ItemMetaMixIn
import no.kh498.util.jackson.serializers.bean.ConfigurationSectionSerializer
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.inventory.meta.ItemMeta


/**
 * @author Elg
 */
class BukkitSerializers(val bukkitModule: BukkitModule) : Serializers.Base() {
    override fun findSerializer(config: SerializationConfig, type: JavaType, beanDesc: BeanDescription): JsonSerializer<*>? {
        return when {
            ItemMeta::class.java.isAssignableFrom(type.rawClass) -> {

                class ItemMetaSerializer : StdSerializer<ItemMeta>(ItemMeta::class.java) {

                    val mapper: ObjectMapper = ObjectMapper().registerModule(BukkitModule(bukkitModule.colorizeStringsByDefault, true))
                    val mapType: MapType = mapper.typeFactory.constructMapType(MutableMap::class.java, String::class.java, Any::class.java)

                    override fun serialize(value: ItemMeta, gen: JsonGenerator, provider: SerializerProvider) {
                        val map = LinkedHashMap<String, Any?>()

                        map[ItemMetaMixIn.TYPE_FIELD] = ItemMetaMixIn.classMap[value::class.java]
                                ?: error("Unknown item meta ${value::class.java}. Perhaps you are using an unsupported version of BukkitUtil?")


                        //the spigot class is weird so we add it as an extra class
                        map.putAll(mapper.convertValue<MutableMap<String, Any?>>(value.spigot(), mapType))
                        map.putAll(mapper.convertValue<MutableMap<String, Any?>>(value, mapType))

                        gen.writeObject(map)
                    }
                }

                return if (bukkitModule.noCustomItemMetaSerialization) null
                else ItemMetaSerializer()
            }
            ConfigurationSection::class.java.isAssignableFrom(type.rawClass) -> {
                ConfigurationSectionSerializer
            }
            else -> null
        }
    }


}
